messaging/app/src/main/java/com/google/firebase/quickstart/fcm/kotlin/MainActivity.kt

```diff
@@ -80,25 +88,46 @@ class MainActivity : AppCompatActivity() {
         binding.logTokenButton.setOnClickListener {
             // Get token
             // [START log_reg_token]
-            Firebase.messaging.getToken().addOnCompleteListener(OnCompleteListener { task ->
-                if (!task.isSuccessful) {
-                    Log.w(TAG, "Fetching FCM registration token failed", task.exception)
-                    return@OnCompleteListener
-                }
-
-                // Get new FCM registration token
-                val token = task.result
-
+            GlobalScope.launch {
+                val token = getAndStoreRegToken()
                 // Log and toast
-                val msg = getString(R.string.msg_token_fmt, token)
-                Log.d(TAG, msg)
-                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
-            })
+                Log.d(TAG, "on click listener: $token")
+                val deviceToken = hashMapOf(
+                    "token" to token,
+                    "timestamp" to FieldValue.serverTimestamp(),
+                )
+
+                // Get user ID from Firebase Auth or your own server
+                Firebase.firestore.collection("fcmTokens").document("myuserid")
+                    .set(deviceToken)
```

Comment:
`.set()` runs asynchronously. And since we're already in a Coroutine scope, maybe we could await for its result too?
```suggestion
                Firebase.firestore.collection("fcmTokens").document("myuserid")
                    .set(deviceToken).await()
```